### PR TITLE
Add IdleConnTimeout to Traefik's http.server settings

### PR DESCRIPTION
**Pull request by bparli**
Without enforcing a timeout Traefik is susceptible to resource leakage, particularly when deployed as a public facing proxy exposed to the Internet.

Make sure IdleTimeout is configurable via the traefik.toml file

Set the default to be 180 seconds

### CODE REVIEW

**Comment by bparli** [comment]
Also, I think this should address #1322

**Comment by timoreimann** [comment]
@bparli thanks for the contribution -- a pretty important one I should say.

Do I see correctly that we need to wait for #1342 before we can merge this?

**Comment by timoreimann** [comment]
I'm also pretty confident this PR will

fix #1322

**Comment by bparli** [comment]
@timoreimann IMO this isn't blocked by #1342; that was just an issue I noticed when testing this.  This setting will be much easier for folks to configure after that issue is fixed though.

**Comment by timoreimann** [comment]
@bparli:

>@timoreimann IMO this isn't blocked by #1342; that was just an issue I noticed when testing this.  This setting will be much easier for folks to configure after that issue is fixed though.

IIUC, configuring is still possible because users may specify (nanoseconds) integers. One thing we have to take into consideration though is that if we use a `time.ParseDuration`-based approach in the future, [specifications missing a unit are going to be rejected](https://play.golang.org/p/md_zcRpk76). What this means is that accepting unit-less specs now will force us to continue providing support for it unless we want to make a breaking change.

WDYT?

**Comment by bparli** [comment]
Thats a fair point, @timoreimann.  My thinking was its already broke for the ProvidersThrottleDuration setting anyway, but you're right, better and cleaner to fix #1342 first.  And it looks like you've already picked that one up.

**Comment by errm** [comment]
LGTM, but waiting for #1342 does make sense ...

**Comment by bparli** [comment]
Sure thing @timoreimann, added some content to toml.md and examples.md.  I also put ProvidersThrottleDuration in the examples.md for good measure.

File: docs/toml.md [hunk 1]
```diff
@@ -67,6 +67,14 @@
 #
 # ProvidersThrottleDuration = "2s"
 
+# IdleTimeout: maximum amount of time an idle (keep-alive) connection will remain idle before closing itself.
+# This is set to enforce closing of stale client connections.
+#
```
Comment (timoreimann) [inline docs/toml.md]: Could you please copy the description of the valid input values [from here](https://github.com/containous/traefik/blame/7d256c9bb98b7786e972cacde02574619bbba21e/docs/toml.md#L13-L15)?

**Comment by bparli** [comment]
Np, good catch to match the ProvidersThrottleDuration language

**Comment by timoreimann** [comment]
Ah sorry, can't merge right away, we need one more pair of maintainer eyes.

@emilevauge could you have a look please?

**Comment by Yggdrasil** [comment]
🥇 Thanks for reporting *and* fixing this @bparli. This issue led to some production downtime for us until we understood what was happening. We've resorted to restarting Traefik twice a week as a workaround. I'll be soooo happy when this is released.

**Comment by bparli** [comment]
My pleasure, happy to contribute

**Comment by Yggdrasil** [comment]
Would it be possible to mention this fix in the Changelog for 1.3.0? I couldn't find it in the RC's and I think it's worth a mention to inform those impacted by the leak.

Below's a sweet example of the difference this fix has made for us on one of our public facing webservers. On this server you can see memory usage climbing from the increasing number of open sockets, until Traefik's restarted. On May 16th around midnight we deployed v1.3.0-rc1, after which memory usage has been stable around 30MB. 😍 

schermafbeelding 2017-05-19 om 16 43 09

**Comment by ldez** [comment]
@Yggdrasil the changelog is fixed. 😃  See #1642.

**Pull request merged**